File: prolog/words/test/words_test.pl
:- use_module(library(plunit)).
:- consult('../src/words.pl').

:- begin_tests(words_suite).
test(reverse_string) :-
    reverse_string("hello", Result1),
    assertion(Result1 == "olleh"),
    reverse_string("a", Result2),
    assertion(Result2 == "a"),
    reverse_string("", Result3),
    assertion(Result3 == "").

test(is_palindrome) :-
    is_palindrome("race car", Result1),
    assertion(Result1 == 1),
    is_palindrome("level", Result2),
    assertion(Result2 == 1),
    is_palindrome("hello", Result3),
    assertion(Result3 == 0),
    is_palindrome("a", Result4),
    assertion(Result4 == 1),
    is_palindrome("", Result5),
    assertion(Result5 == 1).

test(kmp_search) :-
    kmp_search("test", "this is a test", Result1),
    assertion(Result1 == 1),
    kmp_search("not", "this is a test", Result2),
    assertion(Result2 == 0),
    kmp_search("", "any string", Result3),
    assertion(Result3 == 1),
    kmp_search("abc", "abc", Result4),
    assertion(Result4 == 1),
    kmp_search("abc", "ab", Result5),
    assertion(Result5 == 0).

test(lcs) :-
    lcs("AGGTAB", "GXTXAYB", Result1),
    assertion(Result1 == 4),
    lcs("ABC", "ABC", Result2),
    assertion(Result2 == 3),
    lcs("ABC", "DEF", Result3),
    assertion(Result3 == 0),
    lcs("", "ABC", Result4),
    assertion(Result4 == 0),
    lcs("ABC", "", Result5),
    assertion(Result5 == 0).

:- end_tests(words_suite).

File: _experimental/_clojure/words/test/words_test.clj
(ns words-test
  (:require [clojure.test :refer [deftest are run-tests testing]]
            [words :refer [reverse-string is-palindrome kmp-search lcs]]))

(deftest test-reverse-string
  (testing "Reverse String"
    (are [s expected] (= (reverse-string s) expected)
      "hello" "olleh"
      "a" "a"
      "" ""
      "abcde" "edcba")))

(deftest test-is-palindrome
  (testing "Is Palindrome"
    (are [s expected] (= (is-palindrome s) expected)
      "race car" true
      "level" true
      "hello" false
      "a" true
      "" true)))

(deftest test-is-substring
  (testing "Is Substring"
    (are [subst strng expected] (= (kmp-search subst strng) expected)
      "test" "this is a test" true
      "not" "this is a test" false
      "" "any string" true
      "abc" "abc" true
      "abc" "ab" false)))

(deftest test-lcs
  (testing "Longest Common Subsequence"
    (are [str1 str2 expected] (= (lcs str1 str2) expected)
      "AGGTAB" "GXTXAYB" 4
      "ABC" "ABC" 3
      "ABC" "DEF" 0
      "" "ABC" 0
      "ABC" "" 0)))

(defn run-all-tests []
  (run-tests 'words-test))